### PR TITLE
Create container conmon rs

### PR DESCRIPTION
**Pull request by jakecorrenti**
Currently the create functionality of the OCI Runtime Interface is
working at the bare minimum. It only works for rootful execution.

However, the currently lines that integrate Podman with Conmon-rs have
been commented out to prevent breaking main. Additionally, there is a
significant issue with CPU utilization that I need to figure out.

I also moved shared code between `libpod/oci_conmon_linux.go` and
`libpod/oci_conmon_rs_linux.go` into `libpod/conmon_utils.go`.

Additional Notes:
- Arguments need to get passed to either ConmonServerConfig or
  CreateContainerConfig. If an argument does nothing, it is indicative
  of functionality required on the conmon-rs side.
- There is currently no way to configure the environment of the
  conmon-rs process. This will need to be added on the conmon-rs side.
  Similar situation exists with storing extra files dealt with by the
  conmon-rs process.
- [This](https://github.com/containers/podman/blob/b00e65aa9c071428579a55f91a92f3702765ed85/libpod/oci_conmon_linux.go#L1199-L1228) code is going to need some pretty significant adapting
- Need to wire
  [startCommand](https://github.com/containers/podman/blob/b00e65aa9c071428579a55f91a92f3702765ed85/libpod/oci_conmon_linux.go#L1233)
  into conmon-rs as well.
- [readConmonPipeData](https://github.com/containers/podman/blob/b00e65aa9c071428579a55f91a92f3702765ed85/libpod/oci_conmon_linux.go#L1250)
is going to have to be put in conmon-rs, most likely the
`ConmonServerConfig` struct.
- I speculate the CLI sets the old conmon path by default, and this is
  causing issues with finding the conmon-rs path.
- users will need to install conmon-rs using their `get` script to get
  the binary

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -74,6 +74,82 @@ require (
...
+	google.golang.org/genproto v0.0.0-20220304144024-325a89244dc8 // indirect
+	google.golang.org/grpc v1.44.0 // indirect
+	gopkg.in/square/go-jose.v2 v2.5.1 // indirect
+	gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7 // indirect
+	gopkg.in/yaml.v2 v2.4.0 // indirect
+	sigs.k8s.io/yaml v1.3.0 // indirect
+)
+
+require (
+	github.com/containers/conmon-rs v0.0.0-20220630124224-4454b49993de
```
Comment (Luap99) [inline go.mod:151]: I noticed that conmon-rs depends on podman go code, this circular dependency is just asking for trouble in the future. IMO this needs to be fixed before merging: https://github.com/containers/conmon-rs/issues/504

File: go.mod [hunk 2]
```diff
@@ -11,7 +11,7 @@ require (
 	github.com/container-orchestrated-devices/container-device-interface v0.4.0
 	github.com/containernetworking/cni v1.1.1
 	github.com/containernetworking/plugins v1.1.1
-	github.com/containers/buildah v1.26.1-0.20220609225314-e66309ebde8c
+	github.com/containers/buildah v1.26.1
```
Comment (nalind) [inline go.mod:14]: Per offline discussion, use `go get github.com/containers/buildah@v1.26.1-0.20220609225314-e66309ebde8c` to reverse this change, then use `go mod vendor` and `go mod tidy` to sort out the vendored dependencies.

Comment (jakecorrenti) [reply go.mod:14]: Thanks Nalin, that worked!

**Comment by jakecorrenti** [comment]
> @jakecorrenti how much bloat is this adding to the binary? i don't think CI is hitting that check yet due to the build failure. Not that we should worry too much because this is a necessary change... more just curious.

At this point I honestly do not know. I haven't had the chance to test out what I have (I'm assuming it does not work). I'm sure that there will need to be optimizations considering this is almost exactly the same as `libpod/oci_conmon_linux.go`. I can figure it out for you, though.

File: libpod/oci_conmon_rs_linux.go [hunk 3]
```diff
@@ -0,0 +1,753 @@
...
+		logLevel := logrus.WarnLevel
+		if rootless.IsRootless() {
+			logLevel = logrus.InfoLevel
+		}
+		// TODO: This should be a switch - we are not guaranteed that
+		// there are only 2 valid cgroup managers
+		cgroupParent := ctr.CgroupParent()
+		cgroupPath := filepath.Join(ctr.config.CgroupParent, "conmon")
+		Resource := ctr.Spec().Linux.Resources
+		cgroupResources, err := GetLimits(Resource)
```
[the remark this answers is not in the document]
Comment (jakecorrenti) [reply libpod/oci_conmon_rs_linux.go:696]: That's a good point, I will play around with doing something like that.

**Comment by jakecorrenti** [comment]
@cevich Can you take a look at these failing builds?

**Comment by Luap99** [comment]
@jakecorrenti Did you update conmon-rs to the latest main and rebased podman to latets main? We move some code to c/common to remove the circular dependency, see https://github.com/containers/podman/pull/14804#discussion_r913147625

**Comment by jakecorrenti** [comment]
> @jakecorrenti Did you update conmon-rs to the latest main and rebased podman to latets main? We move some code to c/common to remove the circular dependency, see [#14804 (comment)](https://github.com/containers/podman/pull/14804#discussion_r913147625)

I rebased the current branch with main, but I forgot to update the version of conmon-rs. Should fix that failing test now

**Comment by rhatdan** [comment]
@giuseppe Can you take this over?

**Comment by cevich** [comment]
Ref: CI Update for conmon-rs is #15181

**Comment by cdoern** [comment]
I can also take over some of the common rs integration, alot of the broken stuff has to do with rootless cgroup creation if I recall

**Comment by mheon** [comment]
Rootless is a bit broken, but the biggest issues remaining are in conmon-rs itself:

- Exec support
- Ability to pass FDs in
- Checkpoint and Restore
- File logging

**Comment by jakecorrenti** [comment]
I can take on those feature requests you mentioned previously that need work on the conmon-rs side as well.

**Comment by mheon** [comment]
File logging would be a good candidate. We're looking to produce logs formatted identically to the existing logs, which AFAIK is a fairly well-documented format - it's the Kubernetes standard

**Comment by jakecorrenti** [comment]
I'll start working on that end of the week / start of next (currently on vacation)

**Comment by cevich** [comment]
If you think it would help development, I could update #15181 so the status of all conmon-rs tasks is ignored (so everybody's PRs don't turn red).  This would have the side-effect of running our thousands of e2e and system tests with conmon-rs, by commenting out a one-line "ignore" instruction in the YAML.  LMK if you think that would help.  There's no way I'm going to fix #15181 on my own :grin:

**Comment by cevich** [comment]
> Rootless is a bit broken, but the biggest issues remaining are in conmon-rs itself

@mheon and @jakecorrenti FWIW, recent results from my conmon-rs PR (uses latest upstream `main` conmon-rs binary):

* [int podman fedora-36 root host-rs](https://api.cirrus-ci.com/v1/artifact/task/5359255183163392/html/int-podman-fedora-36-root-host-rs.log.html) -- 727 Passed | 976 Failed | 0 Flaked | 0 Pending | 79 Skipped
* [int podman fedora-36 rootless host-rs](https://api.cirrus-ci.com/v1/artifact/task/5499992671518720/html/int-podman-fedora-36-rootless-host-rs.log.html) -- 706 Passed | 901 Failed | 0 Flaked | 0 Pending | 175 Skipped
* [int remote fedora-36 root host-rs](https://api.cirrus-ci.com/v1/artifact/task/5781467648229376/html/int-remote-fedora-36-root-host-rs.log.html) -- 647 Passed | 903 Failed | 0 Flaked | 0 Pending | 218 Skipped

**Comment by mheon** [comment]
Fully expected, given the last of exec/checkpoint

**Pull request closed without merging**